File: lib/java/openlinktoken-core-ai/src/main/java/org/openlinktoken/core/ai/tokentransformer/rotation/RotationMatrixGenerator.java
/* SPDX-License-Identifier: MIT */
package org.openlinktoken.core.ai.tokentransformer.rotation;

import java.nio.charset.StandardCharsets;
import java.security.InvalidKeyException;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;
import java.util.ArrayList;
import java.util.List;

import javax.crypto.Mac;
import javax.crypto.spec.SecretKeySpec;

/**
 * Generates deterministic orthogonal rotation matrices from an initialization vector (IV).
 *
 * <p>The algorithm uses HMAC-SHA256 in counter mode as a PRNG, Box-Muller transform to produce
 * standard-normal values, and Householder QR decomposition for orthonormalization. The
 * resulting columns are normalized by the signs of the QR diagonal to match the persisted
 * PersonMatching rotation-matrix contract.
 *
 * <p>Each returned matrix Q satisfies {@code Q * Q^T = I} and {@code det(Q) = +1}.
 */
public final class RotationMatrixGenerator {

    /**
     * 2^53: used to convert 53-bit integers to uniform doubles.
     * Stored as a long to keep the bit pattern exact.
     */
    private static final long MANTISSA_BITS = 1L << 53;

    private static final double MANTISSA_SCALE = 1.0 / MANTISSA_BITS;

    /** Smallest representable uniform value; avoids log(0) in Box-Muller. */
    private static final double MIN_UNIFORM = MANTISSA_SCALE;

    private static final double TWO_PI = 2.0 * Math.PI;

    private RotationMatrixGenerator() {
    }

    /**
     * Generate a list of deterministic orthogonal rotation matrices from an IV.
     *
     * @param iv            initialization vector string; same IV always produces the same matrices.
     * @param rotationCount number of rotation matrices to generate.
     * @param dimension     number of rows and columns in each matrix (N×N).
     * @return a list of {@code rotationCount} matrices, each a {@code dimension × dimension}
     *         row-major {@code double[][]} where {@code matrix[row][col]} is the element at
     *         (row, col). Each matrix is an orthogonal proper-rotation matrix.
     */
    public static List<double[][]> generate(String iv, int rotationCount, int dimension) {
        return generate(iv, rotationCount, dimension, dimension);
    }

    /**
     * Generate deterministic rotation matrices while retaining only their leading rows.
     *
     * <p>Each full QR matrix is generated before its leading {@code rowCount} rows are copied
     * into the result. This keeps the full-matrix generation and numerical values unchanged
     * while avoiding retention of unused rows.
     *
     * @param iv            initialization vector string; same IV always produces the same matrices.
     * @param rotationCount number of rotation matrices to generate.
     * @param dimension     number of columns and rows in each full matrix (N×N).
     * @param rowCount      number of leading rows to retain in each returned matrix.
     * @return a list of {@code rotationCount} matrices, each a {@code rowCount × dimension}
     *         row-major {@code double[][]}.
     * @throws IllegalArgumentException if {@code rowCount} is not in {@code (0, dimension]}.
     */
    public static List<double[][]> generate(String iv, int rotationCount, int dimension, int rowCount) {
        if (rowCount <= 0 || rowCount > dimension) {
            throw new IllegalArgumentException(
                    "rowCount must be greater than zero and no greater than dimension");
        }

        byte[] keyMaterial = sha256(iv.getBytes(StandardCharsets.UTF_8));
        List<double[][]> matrices = new ArrayList<>(rotationCount);
        for (int r = 0; r < rotationCount; r++) {
            double[][] matrix = generateOne(keyMaterial, r, dimension);
            matrices.add(rowCount == dimension ? matrix : copyLeadingRows(matrix, rowCount));
        }
        return matrices;
    }

    /**
     * Copy the leading rows of a matrix without sharing row arrays.
     */
    private static double[][] copyLeadingRows(double[][] matrix, int rowCount) {
        double[][] leadingRows = new double[rowCount][];
        for (int row = 0; row < rowCount; row++) {
            leadingRows[row] = matrix[row].clone();
        }
        return leadingRows;
    }

    /**
     * Generate one matrix by deriving Gaussian samples from the IV and orthonormalizing them.
     */
    private static double[][] generateOne(byte[] keyMaterial, int rotationIndex, int n) {
        int pairsPerCol = (n + 1) / 2;

        // Build row-major raw matrix filled column-by-column via Box-Muller.
        double[][] raw = new double[n][n];
        for (int col = 0; col < n; col++) {
            int offset = 0;
            for (int pair = 0; pair < pairsPerCol; pair++) {
                long counter = ((long) (rotationIndex * n + col)) * pairsPerCol + pair;
                byte[] h = hmacSha256(keyMaterial, longToBytes(counter));
                double u1 = Math.max(extractUniform(h, 0), MIN_UNIFORM);
                double u2 = extractUniform(h, 8);
                double rVal = Math.sqrt(-2.0 * Math.log(u1));
                double theta = TWO_PI * u2;
                double z0 = rVal * Math.cos(theta);
                double z1 = rVal * Math.sin(theta);
                raw[offset][col] = z0;
                offset++;
                if (offset < n) {
                    raw[offset][col] = z1;
                    offset++;
                }
            }
        }

        double[][] q = householderQr(raw, n);

        // Ensure det(Q) = +1 (proper rotation, no reflection).
        if (computeDetSign(q, n) < 0) {
            for (int row = 0; row < n; row++) {
                q[row][n - 1] = -q[row][n - 1];
            }
        }

        return q;
    }

    /**
     * Compute the orthogonal factor of a matrix using Householder QR decomposition.
     */
    private static double[][] householderQr(double[][] raw, int n) {
        double[][] q = identityMatrix(n);
        double[][] r = copyMatrix(raw, n);

        for (int col = 0; col < n; col++) {
            double[] reflector = new double[n - col];
            for (int row = col; row < n; row++) {
                reflector[row - col] = r[row][col];
            }

            double norm = vectorNorm(reflector);
            if (norm == 0.0) {
                continue;
            }
            reflector[0] += Math.copySign(norm, reflector[0]);
            applyReflectorFromLeft(r, reflector, col, col, n);
            applyReflectorFromRight(q, reflector, col, n);
        }

        for (int col = 0; col < n; col++) {
            if (r[col][col] < 0.0) {
                for (int row = 0; row < n; row++) {
                    q[row][col] = -q[row][col];
                }
            }
        }
        return q;
    }

    /**
     * Create an identity matrix used as the initial orthogonal factor.
     */
    private static double[][] identityMatrix(int n) {
        double[][] identity = new double[n][n];
        for (int row = 0; row < n; row++) {
            identity[row][row] = 1.0;
        }
        return identity;
    }

    /**
     * Copy a square matrix without sharing its row arrays.
     */
    private static double[][] copyMatrix(double[][] matrix, int n) {
        double[][] copy = new double[n][];
        for (int row = 0; row < n; row++) {
            copy[row] = matrix[row].clone();
        }
        return copy;
    }

    /**
     * Compute a Euclidean norm with scaling to reduce overflow and underflow risk.
     */
    private static double vectorNorm(double[] values) {
        double scale = 0.0;
        double sumOfSquares = 1.0;
        for (double value : values) {
            double absoluteValue = Math.abs(value);
            if (absoluteValue == 0.0) {
                continue;
            }
            if (scale < absoluteValue) {
                sumOfSquares = 1.0 + sumOfSquares * (scale / absoluteValue) * (scale / absoluteValue);
                scale = absoluteValue;
            } else {
                sumOfSquares += (absoluteValue / scale) * (absoluteValue / scale);
            }
        }
        return scale == 0.0 ? 0.0 : scale * Math.sqrt(sumOfSquares);
    }

    /**
     * Apply a Householder reflector to matrix rows from {@code startRow} onward.
     */
    private static void applyReflectorFromLeft(
            double[][] matrix, double[] reflector, int startRow, int startColumn, int n) {
        double squaredNorm = squaredNorm(reflector);
        for (int col = startColumn; col < n; col++) {
            double dotProduct = 0.0;
            for (int row = startRow; row < n; row++) {
                dotProduct += reflector[row - startRow] * matrix[row][col];
            }
            double multiplier = 2.0 * dotProduct / squaredNorm;
            for (int row = startRow; row < n; row++) {
                matrix[row][col] -= multiplier * reflector[row - startRow];
            }
        }
    }

    /**
     * Apply a Householder reflector to matrix columns from {@code startColumn} onward.
     */
    private static void applyReflectorFromRight(double[][] matrix, double[] reflector, int startColumn, int n) {
        double squaredNorm = squaredNorm(reflector);
        for (int row = 0; row < n; row++) {
            double dotProduct = 0.0;
            for (int col = startColumn; col < n; col++) {
                dotProduct += matrix[row][col] * reflector[col - startColumn];
            }
            double multiplier = 2.0 * dotProduct / squaredNorm;
            for (int col = startColumn; col < n; col++) {
                matrix[row][col] -= multiplier * reflector[col - startColumn];
            }
        }
    }

    /**
     * Compute the squared Euclidean norm of a vector.
     */
    private static double squaredNorm(double[] values) {
        double sum = 0.0;
        for (double value : values) {
            sum += value * value;
        }
        return sum;
    }

    /**
     * Extract 53 bits from 8 bytes of HMAC output starting at {@code offset} and scale to [0, 1).
     */
    private static double extractUniform(byte[] hmacBytes, int offset) {
        long value = 0;
        for (int i = 0; i < 8; i++) {
            value = (value << 8) | (hmacBytes[offset + i] & 0xFF);
        }
        return ((value >>> 11) & (MANTISSA_BITS - 1)) * MANTISSA_SCALE;
    }

    /**
     * Return +1 or -1: the sign of det(Q) via Gaussian elimination with partial pivoting.
     */
    private static int computeDetSign(double[][] q, int n) {
        double[][] a = new double[n][];
        for (int i = 0; i < n; i++) {
            a[i] = q[i].clone();
        }
        int sign = 1;
        for (int col = 0; col < n; col++) {
            int maxRow = col;
            for (int row = col + 1; row < n; row++) {
                if (Math.abs(a[row][col]) > Math.abs(a[maxRow][col])) {
                    maxRow = row;
                }
            }
            if (maxRow != col) {
                double[] tmp = a[col];
                a[col] = a[maxRow];
                a[maxRow] = tmp;
                sign = -sign;
            }
            for (int row = col + 1; row < n; row++) {
                double factor = a[row][col] / a[col][col];
                for (int j = col; j < n; j++) {
                    a[row][j] -= factor * a[col][j];
                }
            }
        }
        for (int i = 0; i < n; i++) {
            if (a[i][i] < 0) {
                sign = -sign;
            }
        }
        return sign;
    }

    /**
     * Encode a long as an eight-byte big-endian value for HMAC counter input.
     */
    private static byte[] longToBytes(long value) {
        byte[] bytes = new byte[8];
        for (int i = 7; i >= 0; i--) {
            bytes[i] = (byte) (value & 0xFF);
            value >>= 8;
        }
        return bytes;
    }

    /**
     * Hash input with SHA-256.
     */
    private static byte[] sha256(byte[] data) {
        try {
            return MessageDigest.getInstance("SHA-256").digest(data);
        } catch (NoSuchAlgorithmException e) {
            throw new RuntimeException("SHA-256 not available", e);
        }
    }

    /**
     * Authenticate input with HMAC-SHA256 using the supplied key.
     */
    private static byte[] hmacSha256(byte[] key, byte[] data) {
        try {
            Mac mac = Mac.getInstance("HmacSHA256");
            mac.init(new SecretKeySpec(key, "HmacSHA256"));
            return mac.doFinal(data);
        } catch (NoSuchAlgorithmException | InvalidKeyException e) {
            throw new RuntimeException("HmacSHA256 not available", e);
        }
    }
}

File: lib/python/openlinktoken-core-ai/src/main/openlinktoken/core/ai/tokentransformer/rotation/rotation_matrix_generator.py
# SPDX-License-Identifier: MIT

import hashlib
import math
from typing import List, Optional

import numpy as np

# 2^53 used to convert 53-bit integers to uniform doubles.
_MANTISSA_BITS = 1 << 53
_MANTISSA_SCALE = 1.0 / _MANTISSA_BITS
# Smallest representable uniform value, avoids log(0) in Box-Muller.
_MIN_UNIFORM = _MANTISSA_SCALE
_TWO_PI = 2.0 * math.pi


def generate(
    iv: str,
    rotation_count: int,
    dimension: int,
    rows: Optional[int] = None,
) -> List[np.ndarray]:
    """Generate a list of deterministic orthogonal rotation matrices from an IV.

    The matrices are derived from the IV using HMAC-SHA256 in counter mode for
    pseudo-random number generation, Box-Muller transform for standard normal
    values, and QR decomposition for orthonormalization. The algorithm is
    fully specified in terms of standard operations and produces bit-exact
    results across Python and Java implementations.

    Args:
        iv: Initialization vector string. Same IV always produces the same matrices.
        rotation_count: Number of rotation matrices to generate.
        dimension: Number of rows and columns in each matrix.
        rows: Optional number of leading rows to retain from each matrix.

    Returns:
        A list of ``rotation_count`` numpy float64 proper-rotation matrices.
        When ``rows`` is provided, each matrix has shape ``(rows, dimension)``.
    """
    if rows is not None and (rows <= 0 or rows > dimension):
        raise ValueError("rows must be in the range [1, dimension].")

    key_material = hashlib.sha256(iv.encode("utf-8")).digest()
    return [_generate_one(key_material, r, dimension, rows) for r in range(rotation_count)]


def _generate_one(
    key_material: bytes,
    rotation_index: int,
    n: int,
    rows: Optional[int] = None,
) -> np.ndarray:
    """Generate a single ``n x n`` proper-rotation matrix."""
    pairs_per_col = (n + 1) // 2
    sample_count = n * pairs_per_col

    # Reuse HMAC's padded key state and vectorize the Box-Muller transform. The
    # resulting digest bytes and QR input remain identical to the scalar path.
    padded_key = key_material.ljust(64, b"\x00")
    inner_pad = bytes(value ^ 0x36 for value in padded_key)
    outer_pad = bytes(value ^ 0x5C for value in padded_key)
    inner_hash = hashlib.sha256(inner_pad)
    outer_hash = hashlib.sha256(outer_pad)
    digest_bytes = bytearray(sample_count * 16)
    counter_base = rotation_index * sample_count

    for sample_index in range(sample_count):
        inner = inner_hash.copy()
        inner.update((counter_base + sample_index).to_bytes(8, "big"))
        outer = outer_hash.copy()
        outer.update(inner.digest())
        digest_start = sample_index * 16
        digest_bytes[digest_start : digest_start + 16] = outer.digest()[:16]

    digest_words = np.frombuffer(bytes(digest_bytes), dtype=">u8").reshape(sample_count, 2)
    uniforms = ((digest_words >> np.uint64(11)) & np.uint64(_MANTISSA_BITS - 1)).astype(np.float64) * _MANTISSA_SCALE
    u1 = np.maximum(uniforms[:, 0], _MIN_UNIFORM)
    u2 = uniforms[:, 1]
    radius = np.sqrt(-2.0 * np.log(u1))
    theta = _TWO_PI * u2

    interleaved = np.empty(sample_count * 2, dtype=np.float64)
    interleaved[0::2] = radius * np.cos(theta)
    interleaved[1::2] = radius * np.sin(theta)
    raw = interleaved.reshape(n, pairs_per_col * 2)[:, :n].T.copy()

    q, r = np.linalg.qr(raw)
    q = q * np.sign(np.diag(r))[np.newaxis, :]
    if np.sign(np.linalg.det(q)) < 0:
        q[:, n - 1] = -q[:, n - 1]
    return q if rows is None else q[:rows, :].copy()


def _extract_uniform(h: bytes, offset: int) -> float:
    """Extract 53 bits from 8 bytes of HMAC output and scale to [0, 1).

    Args:
        h: HMAC-SHA256 digest bytes (at least ``offset + 8`` bytes long).
        offset: Byte offset within ``h`` to start reading from.

    Returns:
        A float in the range [0, 1) derived from the 53 most-significant bits
        of the 64-bit big-endian integer at ``h[offset:offset+8]``.
    """
    value = int.from_bytes(h[offset : offset + 8], "big")
    return ((value >> 11) & (_MANTISSA_BITS - 1)) * _MANTISSA_SCALE
